README.md
# A Web Scraper for Laravel 5

[![Latest Version on Packagist][ico-version]][link-packagist]
[![Software License][ico-license]](LICENSE.md)
[![Build Status][ico-travis]][link-travis]
[![Coverage Status][ico-scrutinizer]][link-scrutinizer]
[![Quality Score][ico-code-quality]][link-code-quality]
[![Total Downloads][ico-downloads]][link-downloads]

This is where your description should go. Try and limit it to a paragraph or two, and maybe throw in a mention of what
PSRs you support to avoid any confusion with users and contributors.

## Install

Via Composer

``` bash
$ composer require jeroenherczeg/hyena
```

Next, you must install the service provider:

``` php
// config/app.php
'providers' => [
    ...
    Jeroenherczeg\Hyena\HyenaServiceProvider::class,
];
```

And add facade:

``` php
// config/app.php
'aliases' => [
    ...
    Jeroenherczeg\Hyena\Facades\Hyena::class,
];
```

## Usage

``` php
$result = Hyena::visit('https://github.com')->extract(['name', 'images']);
$result = Hyena::visit('https://github.com')->extract(['name', 'images'], [
    'min_image_width'    => 50, // optional, minimal width of picture in px
    'min_image_height'   => 50, // optional, minimal height of picture in px
    'min_image_filesize' => 16000, // optional, minimal filesize of picture in bytes
    'limit_images'       => 10  // optional, max count of images to return
]);
```

``` php
[
  'name' => 'Github',
  'images' => [
    'https://avatars1.githubusercontent.com/u/759412?v=3&s=40',
    'https://assets-cdn.github.com/images/spinners/octocat-spinner-128.gif',
    'https://assets-cdn.github.com/images/spinners/octocat-spinner-32.gif'
  ]
]
```

## Change log

Please see [CHANGELOG](CHANGELOG.md) for more information what has changed recently.

## Testing

``` bash
$ composer test
```

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) and [CONDUCT](CONDUCT.md) for details.

## Security

If you discover any security related issues, please email jeroen@herczeg.be instead of using the issue tracker.

## Credits

- [Jeroen Herczeg][link-author]
- [All Contributors][link-contributors]

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.

[ico-version]: https://img.shields.io/packagist/v/jeroenherczeg/hyena.svg?style=flat-square
[ico-license]: https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square
[ico-travis]: https://img.shields.io/travis/jeroenherczeg/hyena/master.svg?style=flat-square
[ico-scrutinizer]: https://img.shields.io/scrutinizer/coverage/g/jeroenherczeg/hyena.svg?style=flat-square
[ico-code-quality]: https://img.shields.io/scrutinizer/g/jeroenherczeg/hyena.svg?style=flat-square
[ico-downloads]: https://img.shields.io/packagist/dt/jeroenherczeg/hyena.svg?style=flat-square

[link-packagist]: https://packagist.org/packages/jeroenherczeg/hyena
[link-travis]: https://travis-ci.org/jeroenherczeg/hyena
[link-scrutinizer]: https://scrutinizer-ci.com/g/jeroenherczeg/hyena/code-structure
[link-code-quality]: https://scrutinizer-ci.com/g/jeroenherczeg/hyena
[link-downloads]: https://packagist.org/packages/jeroenherczeg/hyena
[link-author]: https://github.com/jeroenherczeg
[link-contributors]: ../../contributors


CHANGELOG.md
# Changelog

All Notable changes to `hyena` will be documented in this file.

Updates should follow the [Keep a CHANGELOG](http://keepachangelog.com/) principles.

## NEXT - YYYY-MM-DD

### Added
- Nothing

### Deprecated
- Nothing

### Fixed
- Nothing

### Removed
- Nothing

### Security
- Nothing


src/Hyena.php
<?php
namespace Jeroenherczeg\Hyena;

use GuzzleHttp\Client;
use GuzzleHttp\Exception\ConnectException;

class Hyena
{
    /**
     * Create a new Skeleton Instance
     */
    public function __construct()
    {
        // constructor body
    }

    public function visit($uri)
    {
        try {
            $client = new Client(['base_uri' => $uri]);
            $response = $client->request('GET', '/', [
                'debug'           => false,
                'connect_timeout' => 5,
                'headers'         => [
                    'User-Agent' => 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/52.0.2743.116 Safari/537.36'
                ]
            ]);
        } catch (ConnectException $e) {
            return null;
        }

        return new HyenaResponse($uri, $response);
    }
}


src/HyenaNameFinder.php
<?php
/**
 * Created by PhpStorm.
 * User: artycake
 * Date: 8/17/16
 * Time: 13:14
 */
namespace Jeroenherczeg\Hyena;

use Symfony\Component\DomCrawler\Crawler;

class HyenaNameFinder
{
    /** @var Crawler */
    private $crawler;
    /** @var string */
    private $uri;
    /** @var string */
    private $domain;
    /** @var array */
    private $names = [];

    /**
     * HyenaNameFinder constructor.
     *
     * @param Crawler $crawler
     * @param string $uri
     */
    public function __construct(Crawler $crawler, $uri)
    {
        $this->crawler = $crawler;
        $this->uri = $uri;
        preg_match('/https?:\/\/(www.)?([a-zA-Z0-9.-]+)(?=\.[a-zA-Z{2,3}])/', $this->uri, $parts);
        $this->domain = $parts[2];
    }

    /**
     * Return name parsed from website
     *
     * @return string
     */
    public function getName()
    {
        if ($name = $this->getNameFromOgSiteName()) {
            return $name;
        }
        if ($name = $this->getNameFromTitle()) {
            return $name;
        }
        $this->collectNameStrings();
        $parts = $this->getMostRelevantKeys();
        if (count($parts) == 0) {
            $parts = preg_split('/[\s_,.-]+/', strtolower($this->domain));
        }

        return $this->createNameFromParts($parts);
    }

    /**
     * Add name to names' array and increase count if repeats
     *
     * @param $name
     */
    private function addName($name)
    {
        $name = str_replace('-', ' ', str_slug($name));
        if (!isset($this->names[$name])) {
            $this->names[$name] = ['count' => 0, 'name' => strtolower($name)];
        }
        $this->names[$name]['count']++;
    }

    /**
     * Look for domain name inside title (common case)
     * in case of success return part of title similar to domain name
     * otherwise add title to name arrays and return null
     *
     * @param string $domain
     * @param string $title
     * @return null|string
     */
    private function extractNameFromTitle($domain, $title)
    {
        $pattern = '/[^a-z0-9]+/';
        $compressedDomain = preg_replace($pattern, '', strtolower($domain));
        $compressedTitle = preg_replace($pattern, '', strtolower($title));
        // Title contains no letters or numbers
        if (!$compressedTitle) {
            return null;
        }
        // Title contains only domain name
        if (strpos($compressedDomain, $compressedTitle) !== false) {
            return $this->clearName($title);
        }
        /** @var bool $fromStart - if title starts with domain name */
        $fromStart = strpos($compressedTitle, $compressedDomain) !== 0;
        while (strpos($title, ' ') !== false && $compressedDomain != $compressedTitle) {
            if ($fromStart) {
                $title = substr($title, strpos($title, ' '));
            } else {
                $title = substr($title, 0, strrpos($title, ' '));
            }
            $title = $this->clearName($title);
            $compressedTitle = preg_replace($pattern, '', strtolower($title));
            if (strpos($compressedTitle, $compressedDomain) === 0) {
                $fromStart = false;
            }
        }
        // Remaining part of title equals domain name
        if ($compressedDomain == $compressedTitle) {
            $clearTitle = $this->clearName($title);
            if (strtolower($clearTitle) === $clearTitle && strpos('.', $clearTitle) === false) {
                return $this->createNameFromParts(explode(' ', $clearTitle));
            }

            return $clearTitle;
        }
        // Title contains domain name with domain zone
        if ($fromStart == false) {
            if (strpos($title, '.') !== false) {
                return $this->clearName($title);
            }
        }
        // Domain consists of multiple parts, i.e. subdomains
        // Try to find part of title equals each of part of domain name
        if (strpos($domain, '.') !== false) {
            $domainParts = explode('.', $domain);
            $resultTitle = null;
            while (count($domainParts) && $resultTitle == null) {
                $resultTitle = $this->extractNameFromTitle(array_shift($domainParts), $title);
            }
            if ($resultTitle) {
                return $resultTitle;
            }
        }

        return null;
    }

    /**
     * Compare all names' strings and return most common words
     *
     * @return array
     */
    private function getMostRelevantKeys()
    {
        usort($this->names, function ($a, $b) {
            if ($a['count'] == $b['count']) {
                return 0;
            }

            return $a['count'] > $b['count'] ? -1 : 1;
        });
        $keys = [];
        $keysByName = [];
        if (count($this->names)) {
            foreach ($this->names as $name) {
                $stringName = $name['name'];
                if (!isset($keysByName[$stringName])) {
                    $keysByName[$stringName] = [];
                }
                $keysByName[$stringName] = preg_split('/[\s_,.-]+/', strtolower($stringName));
                foreach ($this->names as $nameToCompare) {
                    $_keys = preg_split('/[\s_,.-]+/', strtolower($nameToCompare['name']));
                    $newKeys = array_intersect($keysByName[$stringName], $_keys);
                    if (count($newKeys)) {
                        $keysByName[$stringName] = $newKeys;
                    }
                }
            }
            usort($keysByName, function ($a, $b) {
                return count($a) > count($b) ? -1 : 1;
            });
            $resultKeys = $keysByName[0];
            foreach ($keysByName as $_keys) {
                $newResultKeys = array_intersect($resultKeys, $_keys);
                if (count($newResultKeys)) {
                    $resultKeys = $newResultKeys;
                }
            }
            $keys = $resultKeys;
        }

        return array_unique($keys);
    }

    /**
     * Look for "og:site_name" meta tag inside page
     * Most relevant name of site defined by owner
     * Return null in case of tag is missing or contains empty value
     *
     * @return null|string
     */
    private function getNameFromOgSiteName()
    {
        $ogSiteNameNode = $this->crawler->filter('[property="og:site_name"]');
        if (iterator_count($ogSiteNameNode)) {
            $nameFromOgTag = $this->clearName($ogSiteNameNode->attr('content'));
            if ($nameFromOgTag) {
                return $nameFromOgTag;
            }
        }

        return null;
    }

    /**
     * Trim all non-significant characters from given string
     *
     * @param string $name
     * @return string
     */
    private function clearName($name)
    {
        $convertedName = urlencode($name);
        // spaces
        $convertedName = str_replace(
            ['%7F', '%81', '%81', '%C5%8D', '%8F', '%C2%90', '%C2%A0', '%0A', '%09'],
            '+',
            $convertedName
        );
        // hyphen
        $convertedName = str_replace(['%E2%80%93'], '-', $convertedName);
        // vertical bar
        $convertedName = str_replace(['%C7%80'], '%7C', $convertedName);
        $name = urldecode($convertedName);
        $clearName = trim($name, ' -_|&:.,•»');

        return $clearName;
    }

    /**
     * Look for title tag
     * Convert title value into latin symbols
     * Try to extract site name from title
     *
     * @return null|string
     */
    private function getNameFromTitle()
    {
        $titleNode = $this->crawler->filter('title');
        if (count($titleNode)) {
            $title = trim($titleNode->text());
            $title = str_replace('-', ' ', str_slug($title));
            $extractedName = $this->extractNameFromTitle($this->domain, $title);
            if ($extractedName) {
                return $extractedName;
            }
            $this->addName($title);
        }
    }

    /**
     * Collect values of all meaningful tags on the page
     * Add found values to names' array
     */
    private function collectNameStrings()
    {
        // Value from H1 tag
        $hNode = $this->crawler->filter('h1');
        if (count($hNode)) {
            foreach ($hNode as $content) {
                $crawler = new Crawler($content);
                $text = trim($crawler->text());
                $this->addName($text);
            }
        }
        // Title or content from link to main page
        $mainPageLinksNode = $this->crawler->filter('a[href="/"], a[href="' . $this->uri . '"], a[href="' . $this->uri . '/"]');
        if (count($mainPageLinksNode)) {
            foreach ($mainPageLinksNode as $content) {
                $linkNode = new Crawler($content);
                $text = trim($linkNode->attr('title'));
                $this->addName($text);
                $text = trim($linkNode->text());
                $this->addName($text);
            }
        }
        //Title or alt from images with logo image
        $logoNode = $this->crawler->filter('.logo img');
        if (count($logoNode)) {
            if ($nameFromLogo = trim($logoNode->attr('title'))) {
                $this->addName($nameFromLogo);
            }
            if ($nameFromLogo = trim($logoNode->attr('alt'))) {
                $this->addName($nameFromLogo);
            }
        }
        //Title or alt from images with logo image
        $logoNode = $this->crawler->filterXPath('.//*[@src[contains(.,\'logo\')]]');
        if (count($logoNode)) {
            if ($nameFromLogo = trim($logoNode->attr('title'))) {
                $this->addName($nameFromLogo);
            }
            if ($nameFromLogo = trim($logoNode->attr('alt'))) {
                $this->addName($nameFromLogo);
            }
        }
        // Value from copyright meta tag
        $copyrightNode = $this->crawler->filterXPath('.//meta[@name[contains(.,\'copyright\')]]');
        if (iterator_count($copyrightNode)) {
            // Clear from copyright sign, year and "By whom"
            $nameFromCopyright = trim(preg_replace('/(©|\d{4}|\..*|\sby.*|copyright)/i', '', $copyrightNode->attr('content')));
            if ($nameFromCopyright) {
                $this->addName($nameFromCopyright);
            }
        }
        // Value from copyright string, commonly in the end of the page
        $copyrightNode = $this->crawler->filterXPath('.//*[text()[contains(.,\'©\')]]');
        if (iterator_count($copyrightNode)) {
            // Clear from copyright sign, year and "By whom"
            $nameFromCopyright = trim(preg_replace('/(©|\d{4}|\..*|\sby.*|copyright)/i', '', $copyrightNode->text()));
            if ($nameFromCopyright) {
                $this->addName($nameFromCopyright);
            }
        }
    }

    /**
     * Convert each word in array to ucfirst
     * Merge words in string trim all non-significant characters
     *
     * @param array $parts
     * @return string
     */
    private function createNameFromParts(array $parts)
    {
        $name = implode(' ', array_map(function ($part) {
            return ucfirst($part);
        }, $parts));

        return $this->clearName($name);
    }
}


src/HyenaImageFinder.php
<?php
/**
 * Created by PhpStorm.
 * User: artycake
 * Date: 8/18/16
 * Time: 10:38
 */
namespace Jeroenherczeg\Hyena;

use Symfony\Component\DomCrawler\Crawler;

class HyenaImageFinder
{
    /** @var Crawler */
    private $crawler;
    /** @var string */
    private $uri;

    /**
     * HyenaImageFinder constructor.
     *
     * @param Crawler $crawler
     * @param string $uri
     */
    public function __construct($crawler, $uri)
    {
        $this->crawler = $crawler;
        $this->uri = $uri;
    }

    /**
     * Look for img tags, check images by links and return links if images match options
     *
     * @param array $options
     * @return array
     */
    public function getImages(array $options)
    {
        $images = [];
        $imageNodes = $this->crawler->filter('[property="og:image"]');
        foreach ($imageNodes as $content) {
            if (count($images) > $options['limit_images']) {
                return $images;
            }
            $imageNode = new Crawler($content);
            $src = $imageNode->attr('content');
            if (!$src) {
                continue;
            }
            if ($this->checkImage($src, $options)) {
                $images[] = $src;
            }
        }
        $imageNodes = $this->crawler->filter('img');
        foreach ($imageNodes as $content) {
            if (count($images) > $options['limit_images']) {
                return $images;
            }
            $imageNode = new Crawler($content);
            $src = $imageNode->attr('src');
            if (!$src) {
                continue;
            }
            if ($this->checkImage($src, $options)) {
                $images[] = $src;
            }
        }

        return $images;
    }

    private function checkImage($src, $options)
    {
        if (strpos($src, 'http') !== 0) {
            $src = trim($this->uri, '/') . '/' . trim($src, '/');
            $src = strtolower($src);
        }
        $image = new \Imagick();
        try {
            $image->pingImage($src);
            if (
                $image->getImageWidth() < $options['min_image_width'] ||
                $image->getImageHeight() < $options['min_image_height'] ||
                $image->getImageSize() < $options['min_image_filesize']
            ) {
                return false;
            }
        } catch (\Exception $e) {
            echo $e->getMessage() . "\n";

            return false;
        }

        return true;
    }
}


src/HyenaResponse.php
<?php
/**
 * Created by PhpStorm.
 * User: artycake
 * Date: 8/15/16
 * Time: 15:37
 */
namespace Jeroenherczeg\Hyena;

use Psr\Http\Message\ResponseInterface;
use Symfony\Component\DomCrawler\Crawler;

class HyenaResponse
{
    /** @var Crawler */
    private $crawler;
    /** @var string */
    private $uri;
    /** @var array */
    private $availableFields = ['name', 'images'];
    private $defaultOptions = [
        'min_image_width'    => 50,
        'min_image_height'   => 50,
        'min_image_filesize' => 16000,
        'limit_images'       => 10
    ];
    private $currentOptions = [];

    /**
     * HyenaResponse constructor.
     *
     * @param string $uri
     * @param ResponseInterface $response
     */
    public function __construct($uri, ResponseInterface $response)
    {
        $content = $response->getBody()->getContents();
        $this->crawler = new Crawler($content);
        $this->uri = $uri;
    }

    /**
     * Try to extract requested data from given uri
     *
     * @param array $fields
     * @param array $options
     * @return array
     * @throws HyenaParamsExtension
     */
    public function extract(array $fields, array $options = [])
    {
        $result = [];
        $this->currentOptions = array_merge($this->defaultOptions, $options);
        foreach ($fields as $field) {
            if (!is_string($field)) {
                throw new HyenaParamsExtension('Field name should be a string. ' . gettype($field) . ' is given.');
            }
            if (!in_array($field, $this->availableFields)) {
                throw new HyenaParamsExtension('Field "' . $field . '" not available for request. Available fields: ' . implode(', ', $this->availableFields));
            }
            $result[$field] = $this->getFieldFromResponse($field);
        }

        return $result;
    }

    /**
     * Return value for requested field
     *
     * @param string $field
     * @return array|null|string
     */
    private function getFieldFromResponse($field)
    {
        switch ($field) {
            case 'name':
                return $this->getSiteName();
                break;
            case 'images':
                return $this->getSiteImages();
                break;
            default:
                return null;
        }
    }

    /**
     * Collect and return images matching options
     *
     * @return array
     */
    private function getSiteImages()
    {
        $nameFinder = new HyenaImageFinder($this->crawler, $this->uri);

        return $nameFinder->getImages($this->currentOptions);
    }

    /**
     * Find and return name of website
     *
     * @return string
     */
    private function getSiteName()
    {
        $nameFinder = new HyenaNameFinder($this->crawler, $this->uri);

        return $nameFinder->getName();
    }
}
